### PR TITLE
Change underlying dataformat of ScmDataFrame

### PR DESCRIPTION
**Pull request by lewisjared**
# Pull request

Please confirm that this pull request has done the following:

- [x] Tests added
- [x] Documentation added (where applicable)
- [x] Example added (either to an existing notebook or as a new notebook, where applicable)
- [x] Description in ``CHANGELOG.rst`` added

## Adding to CHANGELOG.rst

Please add a single line in the changelog notes similar to one of the following:

```
- (`#XX <https://github.com/openclimatedata/openscm/pull/XX>`_) Added feature which does something
- (`#XX <https://github.com/openclimatedata/openscm/pull/XX>`_) Fixed bug identified in `#YY <https://github.com/openclimatedata/openscm/issues/YY>`_
```

### CODE REVIEW

File: openscm/highlevel.py [hunk 1]
```diff
@@ -4,7 +4,7 @@
 data, easy setting of parameters and stochastic ensemble runs.
 """
 from .core import Core
-from .scmdataframebase import ScmDataFrameBase, df_append
+from .scmdataframebase import ScmDataFrameBase, df_append  # pylint: disable=unused-import
```
Comment (znicholls) [inline openscm/highlevel.py]: what is this wizadry?

File: openscm/highlevel.py [hunk 2]
```diff
@@ -20,7 +20,9 @@ class OpenSCM(Core):
 class ScmDataFrame(ScmDataFrameBase):
     """OpenSCM's custom DataFrame implementation.
 
-    The ScmDataFrame wraps around `pyam`'s IamDataFrame, which itself wraps around Pandas.
+    The ScmDataFrame implements a subset of the functionality provided by `pyam`'s
+    IamDataFrame, but is focused on the providing a performant way of storing
+    time series data and the metadata associated with those time series.
```
Comment (znicholls) [inline openscm/highlevel.py]: For users who wish to take advantage of all of Pyam's functionality, please cast your data frame to an IamDataFrame first with `to_iamdataframe()`. Note: this operation can be relatively computationally expensive for large data sets.

File: openscm/highlevel.py [hunk 3]
```diff
@@ -20,7 +20,9 @@ class OpenSCM(Core):
...
 
     The ScmDataFrame provides a number of diagnostic features (including validation
     of data, completeness of variables provided, running of simple climate models)
```
Comment (znicholls) [inline openscm/highlevel.py]: Can probably delete these lines now that we aren't subclassing `IamDataFrame`

File: tests/conftest.py [hunk 4]
```diff
@@ -32,7 +34,73 @@
...
+    yield IamDataFrame(TEST_DF)
+
+
+@pytest.fixture(
+    scope="function",
+    params=[
+        {"data": TEST_DF},
+        {"data": IamDataFrame(TEST_DF).data},
+        {"data": IamDataFrame(TEST_DF).timeseries()},
+        # {'data': TEST_TS, 'columns': {
```
Comment (znicholls) [inline tests/conftest.py]: do we want this test in or out?

File: openscm/scmdataframebase.py [hunk 5]
```diff
@@ -1,34 +1,504 @@
...
+    datetime_match,
+    hour_match,
+    pattern_match,
+    to_int,
+)
+
+logger = getLogger(__name__)
+
+
+def read_pandas(fname, *args, **kwargs):
```
Comment (znicholls) [inline openscm/scmdataframebase.py]: do we need to copy this function out of pyam?

File: openscm/scmdataframebase.py [hunk 6]
```diff
@@ -1,34 +1,504 @@
...
+        df = pd.read_csv(fname, *args, **kwargs)
+    else:
+        xl = pd.ExcelFile(fname)
+        if len(xl.sheet_names) > 1 and "sheet_name" not in kwargs:
+            kwargs["sheet_name"] = "data"
+        df = pd.read_excel(fname, *args, **kwargs)
+    return df
+
+
+def read_files(fnames, *args, **kwargs):
```
Comment (znicholls) [inline openscm/scmdataframebase.py]: do we need to copy this function out of pyam?

File: openscm/scmdataframebase.py [hunk 7]
```diff
@@ -1,34 +1,504 @@
...
+    if not isstr(fnames):
+        raise ValueError(
+            "reading multiple files not supported, "
+            "please use `pyam.IamDataFrame.append()`"
+        )
+    logger.info("Reading `{}`".format(fnames))
+    return format_data(read_pandas(fnames, *args, **kwargs))
+
+
+def format_data(df, **kwargs):
```
Comment (znicholls) [inline openscm/scmdataframebase.py]: do we need to copy this entire function out of pyam or is there a way to split `format_data` in pyam so we can not do the reshaping bit here?

File: openscm/scmdataframebase.py [hunk 8]
```diff
@@ -1,34 +1,504 @@
...
+        """Initialize an instance of an ScmDataFrameBase
+
+        Parameters
+        ----------
+        data: pd.DataFrame, np.ndarray or data file
+            A pd.DataFrame or data file with IAMC-format data columns, or a numpy array of timeseries data if `columns` is specified.
+        columns: dict
+            dictionary containing the key-value pair of metadata for the timeseries being passed as a numpy array. The values must be
+            arrays of length 1 or length n where n is the number of timeseries.
+        """
```
Comment (znicholls) [inline openscm/scmdataframebase.py]: where do kwargs end up?

File: openscm/scmdataframebase.py [hunk 9]
```diff
@@ -1,34 +1,504 @@
...
+    def __init__(self, data, columns=None, **kwargs):
+        """Initialize an instance of an ScmDataFrameBase
+
+        Parameters
+        ----------
+        data: pd.DataFrame, np.ndarray or data file
+            A pd.DataFrame or data file with IAMC-format data columns, or a numpy array of timeseries data if `columns` is specified.
+        columns: dict
+            dictionary containing the key-value pair of metadata for the timeseries being passed as a numpy array. The values must be
+            arrays of length 1 or length n where n is the number of timeseries.
```
Comment (znicholls) [inline openscm/scmdataframebase.py]: what does this mean?

File: openscm/scmdataframebase.py [hunk 10]
```diff
@@ -1,34 +1,504 @@
...
+            self._meta.values.T, names=self._meta.columns
+        )
+        return d.T
+
+    @property
+    def meta(self):
+        return self._meta.copy()
+
+    def filter(self, keep=True, inplace=False, **kwargs):
+        """Return a filtered IamDataFrame (i.e., a subset of current data)
```
Comment (znicholls) [inline openscm/scmdataframebase.py]:
```suggestion
        """Return a filtered ScmDataFrame (i.e., a subset of current data)
```

File: openscm/scmdataframebase.py [hunk 11]
```diff
@@ -1,34 +1,504 @@
...
+            df = df.set_index(index.names)
+        self._meta = (
+            pd.merge(df, meta, left_index=True, right_index=True, how="outer")
+                .reset_index()
+                .set_index("index")
+        )
+
+
+class LongIamDataFrame(IamDataFrame):
+    """This baseclass is a custom implementation of the IamDataFrame which handles data which spans longer than which pd.to_datetime
```
Comment (znicholls) [inline openscm/scmdataframebase.py]:
```suggestion
    """This baseclass is a custom implementation of the IamDataFrame which handles datetime data which spans longer than pd.to_datetime
```

File: openscm/scmdataframebase.py (hunk 7 above)
Comment (znicholls) [reply openscm/scmdataframebase.py]: Fair, it's also impossible to sensibly split as `format_data` is such an enormous function in pyam. I'll make a PR into pyam to split it. If we want we could reduce duplication in future.

**Comment by lewisjared** [comment]
@znicholls When/where do the emissions data files get downloaded. The notebook can't seem to find them: 
ValueError: no data file `/home/travis/virtualenv/python3.6.7/lib/python3.6/site-packages/openscm/scenarios/rcp26_emissions.csv` found!

File: tests/conftest.py (hunk 4 above)
Comment (lewisjared) [reply tests/conftest.py]: Made the test pass

**Pull request merged**